artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java

```diff
@@ -1366,17 +1367,32 @@ private static void applyExpiryDelay(Message message, AddressSettings settings)
             message.setExpiration(System.currentTimeMillis() + expirationOverride);
          }
       } else {
-         long minExpiration = settings.getMinExpiryDelay();
-         long maxExpiration = settings.getMaxExpiryDelay();
-
-         if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && (message.getExpiration() == 0 || message.getExpiration() > (System.currentTimeMillis() + maxExpiration))) {
-            message.setExpiration(System.currentTimeMillis() + maxExpiration);
+         // if the incoming message has NO expiration then apply the max if set and if not set then apply the min if set
+         if (message.getExpiration() == 0) {
+            if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && maxExpiration != 0) {
+               message.setExpiration(getExpirationToSet(maxExpiration));
+            } else if (minExpiration != AddressSettings.DEFAULT_MIN_EXPIRY_DELAY && minExpiration != 0) {
+               message.setExpiration(getExpirationToSet(minExpiration));
+            }
```

Comment:
The idea with `&& maxExpiration != 0` was simply to avoid setting the expiration to `0` when it already was `0`. However, I see the issue with the logic that you mention so I'm taking it out. Good catch!

Reply:
I didnt think it should be removed, just otherwise fixed.

Allowing it to call setExpirartion(0) registers the message as having been changed, when it has not been. I think you were correct to avoid that case occurring.

Reply:
Done.